src/main/drivers/timer_def.h

```diff
@@ -265,182 +266,111 @@
-#define DMA2_ST6_STREAM                DMA2_Stream6
-#define DMA2_ST7_STREAM                DMA2_Stream7
+// map to base channel (strip N from channel); works only when channel N exists
+#define DEF_TIM_CH2BCH(timch) CONCAT(B, timch)
+#define BCH_TIM1_CH1N BCH_TIM1_CH1
+#define BCH_TIM1_CH2N BCH_TIM1_CH2
+#define BCH_TIM1_CH3N BCH_TIM1_CH3
+
+#define BCH_TIM8_CH1N BCH_TIM8_CH1
+#define BCH_TIM8_CH2N BCH_TIM8_CH2
+#define BCH_TIM8_CH3N BCH_TIM8_CH3
+
+// fetch record from DEF_TIM_DMA__x table
+#define DEF_TIM_DMA_GET(variant, timch) CONCAT4(DEF_TIM_DMA__, variant, __, DEF_TIM_CH2BCH(timch))
+
+// extract members from definition tupple. Concatenation is used to create macro call
+#define DEF_DMA_GET_DMA_STREAM__D(dma_n, stream_n, chan_n) DMA ## dma_n ## _Stream ## stream_n
+#define DEF_DMA_GET_DMA_STREAM__NONE NULL
+#define DEF_DMA_GET_DMA_HANDLER__D(dma_n, stream_n, chan_n) DMA ## dma_n ## _ST ## stream_n ## _HANDLER
+#define DEF_DMA_GET_DMA_HANDLER__NONE 0
+#define DEF_DMA_GET_DMA_CHAN__D(dma_n, stream_n, chan_n) DMA_Channel_ ## chan_n
+#define DEF_DMA_GET_DMA_CHAN__NONE DMA_Channel_0
+
+// map variant + timer channel (base, no N) to description tupple (or NONE)
+// DMA definition                       D(DMAx, Stream, Channel)
+// when unavailable                     NONE
+#define DEF_TIM_DMA__0__BCH_TIM1_CH1    D(2, 6, 0)
+#define DEF_TIM_DMA__1__BCH_TIM1_CH1    D(2, 1, 6)
+#define DEF_TIM_DMA__2__BCH_TIM1_CH1    D(2, 3, 6)
+#define DEF_TIM_DMA__0__BCH_TIM1_CH2    D(2, 6, 0)
+#define DEF_TIM_DMA__1__BCH_TIM1_CH2    D(2, 2, 6)
+#define DEF_TIM_DMA__0__BCH_TIM1_CH3    D(2, 6, 0)
+#define DEF_TIM_DMA__1__BCH_TIM1_CH3    D(2, 6, 6)
+#define DEF_TIM_DMA__0__BCH_TIM1_CH4    D(2, 4, 6)
+
+#define DEF_TIM_DMA__0__BCH_TIM2_CH1    D(1, 5, 3)
+#define DEF_TIM_DMA__0__BCH_TIM2_CH2    D(1, 6, 3)
+#define DEF_TIM_DMA__0__BCH_TIM2_CH3    D(1, 1, 3)
+#define DEF_TIM_DMA__0__BCH_TIM2_CH4    D(1, 7, 3)
+#define DEF_TIM_DMA__1__BCH_TIM2_CH4    D(1, 6, 3)
+
+#define DEF_TIM_DMA__0__BCH_TIM3_CH1    D(1, 4, 5)
+#define DEF_TIM_DMA__0__BCH_TIM3_CH2    D(1, 5, 5)
+#define DEF_TIM_DMA__0__BCH_TIM3_CH3    D(1, 7, 5)
+#define DEF_TIM_DMA__0__BCH_TIM3_CH4    D(1, 2, 5)
+
+#define DEF_TIM_DMA__0__BCH_TIM4_CH1    D(1, 0, 2)
+#define DEF_TIM_DMA__0__BCH_TIM4_CH2    D(1, 3, 2)
+#define DEF_TIM_DMA__0__BCH_TIM4_CH3    D(1, 7, 2)
+
+#define DEF_TIM_DMA__0__BCH_TIM5_CH1    D(1, 2, 6)
+#define DEF_TIM_DMA__0__BCH_TIM5_CH2    D(1, 4, 6)
+#define DEF_TIM_DMA__0__BCH_TIM5_CH3    D(1, 0, 6)
+#define DEF_TIM_DMA__0__BCH_TIM5_CH4    D(1, 1, 6)
+#define DEF_TIM_DMA__1__BCH_TIM5_CH4    D(1, 3, 6)
+
+#define DEF_TIM_DMA__0__BCH_TIM8_CH1    D(2, 2, 0)
+#define DEF_TIM_DMA__1__BCH_TIM8_CH1    D(2, 2, 7)
+#define DEF_TIM_DMA__0__BCH_TIM8_CH2    D(2, 2, 0)
+#define DEF_TIM_DMA__1__BCH_TIM8_CH2    D(2, 3, 7)
+#define DEF_TIM_DMA__0__BCH_TIM8_CH3    D(2, 2, 0)
+#define DEF_TIM_DMA__1__BCH_TIM8_CH3    D(2, 4, 7)
+#define DEF_TIM_DMA__0__BCH_TIM8_CH4    D(2, 7, 7)
```

Comment:
@hydra: With more macro magic (BOOST_PP or P99) multiple variants may be written on single line, with index checking and no entry for non-dma channels:
```
#define DEF_TIM_DMA__TIM8_CH3 (2, 2, 0)(2, 4, 7)
```
But .. I'm quite sceptical about preprocesor library as dependency - it is great when everything works, but debugging it is challenging (I have big problems grasping evaluation rules).
For now it's probably better to avoid too much magic.